src/thread.rs

```diff
@@ -163,43 +169,73 @@ where
 ///
 /// ```
 /// crossbeam_utils::thread::scope(|scope| {
-///     scope.spawn(|| println!("Exiting scope"));
-///     scope.spawn(|| println!("Running child thread in scope"));
+///     scope.spawn(|_| println!("Exiting scope"));
+///     scope.spawn(|_| println!("Running child thread in scope"));
 /// }).unwrap();
 /// ```
 pub fn scope<'env, F, R>(f: F) -> thread::Result<R>
 where
     F: FnOnce(&Scope<'env>) -> R,
 {
+    let (tx, rx) = mpsc::channel();
     let scope = Scope {
-        joins: RefCell::new(Vec::new()),
+        handles: Arc::new(Mutex::new(Vec::new())),
+        chan: tx,
         _marker: PhantomData,
     };
 
     // Execute the scoped function, but catch any panics.
     let result = panic::catch_unwind(panic::AssertUnwindSafe(|| f(&scope)));
+
+    // Wait until all nested scopes are dropped.
+    drop(scope.chan);
+    let _ = rx.recv();
+
     // Join all remaining spawned threads.
-    let mut panics = scope.join_all();
-
-    if panics.is_empty() {
-        result.map_err(|res| Box::new(vec![res]) as _)
-    } else {
-        if let Err(err) = result {
-            panics.reserve(1);
-            panics.insert(0, err);
-        }
+    let panics: Vec<_> = {
+        let mut handles = scope
+            .handles
+            .lock()
+            .unwrap();
+
+        // Filter handles that haven't been joined, join them, and collect errors.
+        let panics = handles
+            .drain(..)
+            .filter_map(|handle| handle.lock().unwrap().take())
+            .filter_map(|handle| handle.join().err())
+            .collect();
+
+        panics
+    };
 
-        Err(Box::new(panics))
+    // If `f` has panicked, resume unwinding.
+    // If any of the child threads have panicked, return the panic errors.
+    // Otherwise, everything is OK and return the result of `f`.
+    match result {
+        Err(err) => panic::resume_unwind(err),
+        Ok(res) => {
+            if panics.is_empty() {
+                Ok(res)
+            } else {
+                Err(Box::new(panics))
+            }
+        }
     }
 }
 
 pub struct Scope<'env> {
-    /// The list of the thread join jobs.
-    joins: RefCell<Vec<Box<FnBox<thread::Result<()>> + 'env>>>,
-    // !Send + !Sync
-    _marker: PhantomData<*const ()>,
+    /// The list of the thread join handles.
+    handles: Arc<Mutex<Vec<Arc<Mutex<Option<thread::JoinHandle<()>>>>>>>,
```

Comment:
I just wonder--are the nested mutexes necessary?

Reply:
Yes:

* The outer mutex is used by nested scopes to add a new join handle when they spawn a new thread.
* The inner mutex is used by the outermost scope to join all remaining threads and by `ScopedJoinHandle::join()`.